src/ion/liveranges.rs

```diff
@@ -1159,66 +1180,108 @@ impl<'a, F: Function> Env<'a, F> {
+                                if self.pregs[preg.index()].is_stack {
+                                    num_fixed_stack += 1;
+                                    first_stack_slot.get_or_insert(u.slot);
+                                } else {
+                                    requires_reg = true;
+                                    num_fixed_reg += 1;
+                                    first_reg_slot.get_or_insert(u.slot);
+                                }
+                            }
+                            // Maybe this could be supported in this future...
+                            OperandConstraint::Stack => panic!(
+                                "multiple uses of vreg with a Stack constraint are not supported"
+                            ),
                         }
                     }
-                };
 
-                for u in &mut self.ranges[range.index()].uses {
-                    let pos = u.pos;
-                    let slot = u.slot as usize;
-                    fixup_multi_fixed_vregs(
-                        pos,
-                        slot,
-                        &mut u.operand,
-                        &mut self.multi_fixed_reg_fixups,
-                    );
+                    // Fast path if there are no conflicts.
+                    if num_fixed_reg + num_fixed_stack <= 1
+                        && !(requires_reg && num_fixed_stack != 0)
+                    {
+                        continue;
+                    }
+
+                    // We pick one constraint (in order: FixedReg, Reg, FixedStack)
+                    // and then rewrite any incompatible constraints to Any.
+                    // This allows register allocation to succeed and we will
+                    // later insert moves to satisfy the rewritten constraints.
+                    let source_slot = if requires_reg {
+                        first_reg_slot.unwrap()
+                    } else {
+                        first_stack_slot.unwrap()
+                    };
+                    let mut first_preg = None;
+                    for u in uses.iter_mut() {
+                        if let OperandConstraint::FixedReg(preg) = u.operand.constraint() {
+                            let vreg_idx = VRegIndex::new(u.operand.vreg().vreg());
+                            let preg_idx = PRegIndex::new(preg.index());
+                            log::trace!(
+                                "at pos {:?}, vreg {:?} has fixed constraint to preg {:?}",
+                                u.pos,
+                                vreg_idx,
+                                preg_idx
+                            );
+
+                            // FixedStack is incompatible if there are any
+                            // Reg/FixedReg constraints. FixedReg is
+                            // incompatible if there already is a different
+                            // FixedReg constraint.
```

Comment:
A little more to this comment would help understanding I think: "If either condition is true, we edit the constraint below; otherwise, we can skip this edit."